roles/postgres/tasks/main.yml

```diff
@@ -0,0 +1,67 @@
+- name: Install postgres Python library.
+  apt:
+    name: "python3-psycopg2"
+    state: present
+
+- name: Install postgres packages.
+  apt:
+    name:
+      - postgresql
+      - postgresql-contrib
+      - libpq-dev
+    state: present
+
+- name: Ensure PostgreSQL data directory exists.
+  file:
+    path: "{{ postgresql_data_dir }}"
+    owner: "{{ postgresql_user }}"
+    group: "{{ postgresql_group }}"
+    state: directory
+    mode: 0700
+
+- name: Check if postgres is initialised.
+  stat:
+    path: "{{ postgresql_data_dir }}/PG_VERSION"
+  register: pgdata_dir_version
+
+- name: Initialise postgres if not already.
+  command: "{{ postgresql_bin_path }}/initdb -D {{ postgresql_data_dir }}"
+  when: not pgdata_dir_version.stat.exists
+  become: true
+  become_user: "{{ postgresql_user }}"
+  # See: https://github.com/ansible/ansible/issues/16048#issuecomment-229012509
+  vars:
+    ansible_ssh_pipelining: true
+
+- name: Check postgres is started and enabled on boot.
+  service:
+    name: "{{ postgresql_daemon }}"
+    state: "{{ postgresql_service_state }}"
+    enabled: "{{ postgresql_service_enabled }}"
+
+- name: Add postgres users.
+  community.postgresql.postgresql_user:
+    name: "{{ item.name }}"
+    login_user: "{{ item.login_user }}"
+    role_attr_flags: "{{ role_attr_flags }}"
+  with_items: "{{ postgresql_users }}"
+  become: true
+  become_user: "{{ postgresql_user }}"
+  vars:
+    ansible_ssh_pipelining: true
+
+- name: Add postgres databases.
+  community.postgresql.postgresql_db:
+    name: "{{ item.name }}"
+    owner: "{{ item.owner | default(postgresql_user) }}"
+  with_items: "{{ postgresql_databases }}"
+  become: true
+  become_user: "{{ postgresql_user }}"
+  vars:
+    ansible_ssh_pipelining: true
```

Comment:
Is there a way to avoid writing this 5 times, or would that involve a bunch of boilerplate that would make it not worthwhile?

Reply:
Yea, it can
> Pipelining can be enabled globally by setting “pipelining=True” in the ssh section of ansible.cfg
However it has a drawback
> Unfortunately, pipelining is disabled by default because it is incompatible with sudo's requiretty setting (or su, which always requires a tty). This is because of a quirk of the Python interpreter, which enters interactive mode automatically when you pipe in data from a (pseudo) tty.

Reply:
Volcyy mentioned yesterday that "Modern Debian comes with requiretty disabled", so I think we can actually enable this globally to get the benefit everywhere.

At that point the only duplicated code is that of the first two lines. I'm not sure how much of an issue this is, nor how we would go about aliasing the become steps.